Update Kubeadm proxy handling for IPv6

rpothier opened the pull request:
This updates HTTPProxyCheck with brackets around
the ipv6 address to handle adding :port


**What this PR does / why we need it**:
For IPv6 addressing, brackets need to be around the IPv6 address when a :port is added.
This change adds the brackets, and also, during testing, an IPV6 NO_PROXY
was not being handled correctly, so this is fixed also by using
netutil.SetOldTransportDefaults as is how HTTPProxyCIDRCheck
handles it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58397

**Special notes for your reviewer**:

**Release note**:

```release-note-none
```

Commit:
Update Kubeadm proxy handling for IPv6

This updates HTTPProxyCheck with brackets around
the ipv6 address to handle adding :port

rpothier commented:
/sig cluster-lifecycle

rpothier commented:
/area ipv6

jamiehannaford commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -415,14 +415,18 @@ func (hst HTTPProxyCheck) Name() string {
 
-	url := fmt.Sprintf("%s://%s:%d", hst.Proto, hst.Host, hst.Port)
+	urlFmt := "%s://%s:%d"
+	if strings.Contains(hst.Host, ":") {
```
is this a fullproof way to check this URL is IPv6? in other PRs we did something like this:

https://github.com/kubernetes/kubernetes/pull/53484/files#diff-86d5c7040c142720e7e2c9894044c353R227

maybe we could centralise as helper functions

liggitt replied:
Don't build URLs with string formatting. Use net.JoinHostPort, and (&url.URL{Scheme:…, Host:…}).String()

dims replied:
+1 to @liggitt 's suggestion, another example is here - https://github.com/kubernetes/kubernetes/pull/57846/commits/c258d4df84089b08b9cbd37b1dee4b00576a2532

rpothier replied:
Thanks, will update.

jamiehannaford reviewed (commented):
just one nit. code looks good but might also be useful to have someone from @kubernetes/sig-network-pr-reviews take a quick look too.

jamiehannaford commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -415,14 +416,15 @@ func (hst HTTPProxyCheck) Name() string {
 
-	url := fmt.Sprintf("%s://%s:%d", hst.Proto, hst.Host, hst.Port)
+	hostPort := net.JoinHostPort(hst.Host, strconv.Itoa(hst.Port))
+	url := fmt.Sprintf("%s://%s", hst.Proto, hostPort)
```
This still builds URLs with strings, please use `(&url.URL{Scheme: hst.Proto, Host: hostPort}).String()` instead

rpothier replied:
ok, will do

dims commented:
/ok-to-test

kad commented on cmd/kubeadm/app/preflight/checks_test.go:
```diff
@@ -503,38 +503,77 @@ func TestHTTPProxyCIDRCheck(t *testing.T) {
 		if (warning != nil) != rt.expectWarnings {
 			t.Errorf(
-				"failed HTTPProxyCIDRCheck:\n\texpected: %t\n\t  actual: %t (CIDR:%s). Warnings: %v",
+				"failed HTTPProxyCheck:\n\texpected: %t\n\t  actual: %t (CIDR:%s). Warnings: %v",
```
in this line then "CIDR:%s" should become "Host:%s"

pmichali replied:
nit: if you want, you could make the error message more descriptive and code more self-documenting, by added a description field to the test structure and use the comments (above) as the description text.  This way, it is easier to tell which test failed, versus comparing the Host IP to the structs above.

rpothier replied:
Will add both these suggestions.

kad commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -415,19 +416,20 @@ func (hst HTTPProxyCheck) Name() string {
 	}
 
-	proxy, err := http.DefaultTransport.(*http.Transport).Proxy(req)
+	proxy, err := netutil.SetOldTransportDefaults(&http.Transport{}).Proxy(req)
```
Interesting. this line should actually trigger merge conflicts, as it was already changed in 4bd692a3bff

rpothier replied:
I'll rebase to pick up those changes.

kad commented:
/lgtm

jamiehannaford commented:
/test pull-kubernetes-bazel-test

kad commented:
/retest
/area kubeadm

timothysc reviewed (approved):
/approve

xiangpengzhao commented:
@rpothier please rebase :)

rpothier commented:
@xiangpengzhao I rebased it, I changed a bit as the port is no longer passed in, however it still removes the creating the url with the printf and adds UT.

kad commented:
/lgtm

rpothier commented:
/test pull-kubernetes-bazel-test

rpothier commented:
/test pull-kubernetes-e2e-kops-aws
